### PR TITLE
Move API endpoints from the config file to the code

### PR DESCRIPTION
**Pull request by tommydeboer**
Depends on: #117 Error when the config contains (old) unknown properties

This PR is a preparation for the compatibility framework

After this PR you can no longer override the API endpoints in the config file, so running integration tests against a MOLGENIS 8 will cause some tests (involving groups) to fail. 

#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/system tested
- [x] User documentation updated
- [x] Test plan template updated
- [x] Clean commits
- [x] Added Feature/Fix to release notes

### CODE REVIEW

File: mcmd/client/api.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@endpoint
+def group():
+    return 'api/plugin/security/group/'
+
+
+@endpoint
+def member(group_name):
+    return 'api/plugin/security/group/{}/member/'.format(group_name)
```
Comment (dennishendriksen) [inline mcmd/client/api.py]: ```group_name``` should be encoded (something like https://stackoverflow.com/a/6431284)

File: mcmd/commands/delete.py [hunk 2]
```diff
@@ -98,15 +98,15 @@ def _delete_package_contents(args):
 
 def _delete_entity_types_in_package(package_id):
     response = client.get(
-        config.api('rest2') + ResourceType.ENTITY_TYPE.get_entity_id() + '?attrs=id&q=package==' + package_id)
+        api.rest2() + ResourceType.ENTITY_TYPE.get_entity_id() + '?attrs=id&q=package==' + package_id)
```
Comment (dennishendriksen) [inline mcmd/commands/delete.py]: - ```package_id``` should be RSQL encoded (as described in https://github.com/jirutka/rsql-parser)
- ```package_id``` should be encoded (something like https://stackoverflow.com/a/51649797)

File: mcmd/commands/delete.py [hunk 3]
```diff
@@ -98,15 +98,15 @@ def _delete_package_contents(args):
...
     entity_ids = [entity_type['id'] for entity_type in response.json()['items']]
     if len(entity_ids) > 0:
         _delete_rows(ResourceType.ENTITY_TYPE.get_entity_id(), entity_ids)
 
 
 def _delete_packages_in_package(package_id):
     response = client.get(
-        config.api('rest2') + ResourceType.PACKAGE.get_entity_id() + '?attrs=id&q=parent==' + package_id)
+        api.rest2() + ResourceType.PACKAGE.get_entity_id() + '?attrs=id&q=parent==' + package_id)
```
Comment (dennishendriksen) [inline mcmd/commands/delete.py]: see previous comment on encoding

File: mcmd/commands/delete.py [hunk 4]
```diff
@@ -116,11 +116,11 @@ def _delete_group(args):
     if args.force or (not args.force and io.confirm(
             'Are you sure you want to delete group {}?'.format(args.resource))):
         io.start('Deleting group {}'.format(highlight(args.resource)))
-        client.delete(urljoin(config.api('group'), args.resource))
+        client.delete(urljoin(api.group(), args.resource))
 
 
 def _delete_rows(entity_type, rows):
-    url = '{}{}'.format(config.api('rest2'), entity_type)
+    url = '{}{}'.format(api.rest2(), entity_type)
```
Comment (dennishendriksen) [inline mcmd/commands/delete.py]: ```entity_type``` needs to be encoded

File: mcmd/commands/set.py [hunk 5]
```diff
@@ -66,7 +66,7 @@ def set_(args):
     io.start(
         'Updating {} of {} settings to {}'.format(highlight(args.setting), highlight(args.type), highlight(args.value)))
     row = _get_row_id(entity)
-    url = '{}{}/{}/{}'.format(config.api('rest1'), entity, row, args.setting)
+    url = '{}{}/{}/{}'.format(api.rest1(), entity, row, args.setting)
```
Comment (dennishendriksen) [inline mcmd/commands/set.py]: URL components should be encoded

File: mcmd/utils/principals.py [hunk 6]
```diff
@@ -30,13 +30,13 @@ def principal_exists(principal_name, principal_type):
 
 def user_exists(username):
     log.debug('Checking if user %s exists' % username)
-    response = get(config.api('rest2') + 'sys_sec_User?q=username==' + username)
+    response = get(api.rest2() + 'sys_sec_User?q=username==' + username)
```
Comment (dennishendriksen) [inline mcmd/utils/principals.py]: - ```username``` should be RSQL encoded
- ```username``` should be URL encoded

File: mcmd/utils/principals.py [hunk 7]
```diff
@@ -30,13 +30,13 @@ def principal_exists(principal_name, principal_type):
...
     return int(response.json()['total']) > 0
 
 
 def role_exists(rolename):
     log.debug('Checking if role %s exists' % rolename)
-    response = get(config.api('rest2') + 'sys_sec_Role?q=name==' + rolename.upper())
+    response = get(api.rest2() + 'sys_sec_Role?q=name==' + rolename.upper())
```
Comment (dennishendriksen) [inline mcmd/utils/principals.py]: ```rolename.upper()``` should be RSQL encoded
```rolename.upper()``` should be URL encoded

File: mcmd/utils/resources.py [hunk 8]
```diff
@@ -53,14 +53,14 @@ def detect_resource_type(resource_id, types: List[ResourceType]):
 def resource_exists(resource_id, resource_type):
     log.debug('Checking if %s %s exists' % (resource_type.get_label(), resource_id))
     query = '?q={}=={}'.format(resource_type.get_identifying_attribute(), resource_id)
```
Comment (dennishendriksen) [inline mcmd/utils/resources.py]: encoding issues (see previous comments)

File: mcmd/utils/resources.py [hunk 9]
```diff
@@ -53,14 +53,14 @@ def detect_resource_type(resource_id, types: List[ResourceType]):
...
-    response = get(config.api('rest2') + resource_type.get_entity_id() + query)
+    response = get(api.rest2() + resource_type.get_entity_id() + query)
```
Comment (dennishendriksen) [inline mcmd/utils/resources.py]: ```resource_type.get_entity_id()``` should be encoded

File: mcmd/utils/resources.py [hunk 10]
```diff
@@ -53,14 +53,14 @@ def detect_resource_type(resource_id, types: List[ResourceType]):
...
     return int(response.json()['total']) > 0
 
 
 def one_resource_exists(resources, resource_type):
     log.debug('Checking if one of [{}] exists in [{}]'.format(','.join(resources), resource_type.get_label()))
     query = '?q={}=in=({})'.format(resource_type.get_identifying_attribute(), ','.join(resources))
```
Comment (dennishendriksen) [inline mcmd/utils/resources.py]: encoded issues (see previous comments)

File: mcmd/utils/resources.py [hunk 11]
```diff
@@ -53,14 +53,14 @@ def detect_resource_type(resource_id, types: List[ResourceType]):
...
-    response = get(config.api('rest2') + resource_type.get_entity_id() + query)
+    response = get(api.rest2() + resource_type.get_entity_id() + query)
```
Comment (dennishendriksen) [inline mcmd/utils/resources.py]: encoding issues (see previous comments)

File: tests/unit/version/molgenis_version_test.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
+import unittest
+
+import pytest
+
+from mcmd.utils.errors import McmdError
+from mcmd.version import molgenis_version
+
+
+@pytest.mark.unit
+class MolgenisVersionTest(unittest.TestCase):
```
Comment (dennishendriksen) [inline tests/unit/version/molgenis_version_test.py]: what is the expected behavior for versions in other molgenis app repositories such as 7.0.0-BBMRI-ERIC?

**Comment by dennishendriksen** [comment]
@tommydeboer could you address the comment of @marikaris?

**Comment by tommydeboer** [comment]
@dennishendriksen @marikaris
>Do we really want to move this from the config? I thought we found it elegant that when the name of an API in molgenis is changed, it can still be used, unless functionality of the API was changed of course.

Moving the location of these strings to the code makes it easier to switch between them based on versions (see #119). If we would want to do this in the config we would have to invent new keys for each endpoint (`group7` and `group8` for example) and add extra code for figuring out which key to use when. This PR is based on the idea that APIs won't constantly change names or move to another location. Even if that happens, it's a matter of adding one method in the code and doing a patch release. I have thought about also making it possible to override the endpoints in the config, but that seems like another feature of which I don't think it's worth the cost to build it.

**Comment by tommydeboer** [comment]
@dennishendriksen  
>I suspect most encoding issues could be prevented by using the MOLGENIS Python API client.

We're not using the Python client (yet), because:
1. There are a lot of other endpoints being used that are not the REST API
2. The Python client uses REST API v1 which will be deprecated
3. Not using it gives us more flexibility of when and how to authenticate
4. Not using it gives us a generalised way to handle _all_ requests to MOLGENIS (+ error handling)

I think we can fix the encoding issues in the `molgenis_client` module.

File: tests/unit/version/molgenis_version_test.py (hunk 12 above)
Comment (tommydeboer) [reply tests/unit/version/molgenis_version_test.py]: See the second to last test case: 

```python
def test_extract_version_number_multi():
        version = molgenis_version._extract_version_number('7.2.3-APP-1.0.0')
        assert version == '7.2.3'
```

**Comment by tommydeboer** [comment]
@dennishendriksen @marikaris
Concerning the escaping of RSQL values: I couldn't find a (reliable looking) Python library that does this. I've added an issue for this problem (#125) and propose to fix this another time since it is not introduced by this PR.

**Comment by tommydeboer** [comment]
@dennishendriksen 
I've added an issue for using the Python REST Client (https://github.com/molgenis/molgenis-tools-commander/issues/126)

File: mcmd/commands/make.py [hunk 13]
```diff
@@ -36,13 +36,16 @@ def make(args):
...
-    url = config.api('member').format(group_name)
-    post(url, {'username': args.user, 'roleName': args.role.upper()})
+    url = api.member(group_name)
+    post(url, data={'username': args.user, 'roleName': args.role.upper()})
 
 
 def _find_group(role):
     io.debug('Fetching groups')
-    groups = get(config.api('rest2') + 'sys_sec_Group?attrs=name')
+    groups = get(api.rest2('sys_sec_Group'),
```
Comment (marikaris) [inline mcmd/commands/make.py]: Do we want this table name hardcoded here?

File: mcmd/commands/set.py [hunk 14]
```diff
@@ -65,15 +65,17 @@ def set_(args):
...
+    url = api.rest1('{}/{}/{}'.format(entity, row, args.setting))
     put(url, json.dumps(args.value))
 
 
 def _get_settings():
-    entity = 'sys_md_EntityType'
-    query = 'q=extends==sys_set_settings&attrs=~id'
-    molgenis_settings = _quick_get(entity, query)
+    molgenis_settings = get(api.rest2('sys_md_EntityType'),
+                            params={
```
Comment (marikaris) [inline mcmd/commands/set.py]: Do we want a hardcoded table name?

File: mcmd/utils/principals.py [hunk 15]
```diff
@@ -30,13 +30,20 @@ def principal_exists(principal_name, principal_type):
...
+                       'q': 'username==' + username
+                   })
+
     return int(response.json()['total']) > 0
 
 
 def role_exists(rolename):
     log.debug('Checking if role %s exists' % rolename)
-    response = get(config.api('rest2') + 'sys_sec_Role?q=name==' + rolename.upper())
+    response = get(api.rest2('sys_sec_Role'),
```
Comment (marikaris) [inline mcmd/utils/principals.py]: Hardcoded tablename

File: mcmd/utils/principals.py [hunk 16]
```diff
@@ -30,13 +30,20 @@ def principal_exists(principal_name, principal_type):
 
 def user_exists(username):
     log.debug('Checking if user %s exists' % username)
-    response = get(config.api('rest2') + 'sys_sec_User?q=username==' + username)
+    response = get(api.rest2('sys_sec_User'),
+                   params={
```
Comment (marikaris) [inline mcmd/utils/principals.py]: hardcoded tablename

**Pull request merged**